addons/l10n_vn_edi_viettel/models/account_move.py

```diff
@@ -0,0 +1,766 @@
+        except (RequestException, ValueError) as err:
+            raise UserError(_('Something went wrong, please try again later: %s', err))
+
+    def _l10n_vn_edi_get_access_token(self):
+        """ Return an access token to be used to contact the API. Either take a valid stored one or get a new one. """
+        self.ensure_one()
+        # First, check if we have a token stored and if it is still valid.
+        if self.company_id.l10n_vn_edi_token and self.company_id.l10n_vn_edi_token_expiry > datetime.now():
+            return self.company_id.l10n_vn_edi_token
+
+        data = {'username': self.company_id.l10n_vn_edi_username, 'password': self.company_id.l10n_vn_edi_password}
+        request_response, error_message = self._l10n_vn_edi_send_request(
+            method='POST',
+            endpoint='/auth/login',  # The / here will cause the route to be added to the base domain of the API
+            json_data=data
+        )
+        if error_message:
+            raise UserError(error_message)
+        if 'access_token' not in request_response:  # Just in case something else go wrong and it's missing the token
+            raise UserError(_('Connection to the API failed, please try again later.'))
+
+        access_token = request_response['access_token']
+        access_token_expiry = datetime.now() + relativedelta(seconds=request_response['expires_in'])
+
+        # Tokens are valid for 5 minutes. Storing it helps reduce api calls and speed up things a little bit.
+        self.company_id.write({
+            'l10n_vn_edi_token': access_token,
+            'l10n_vn_edi_token_expiry': access_token_expiry,
+        })
+
+        return request_response['access_token']
+
+    # -------------------------------------------------------------------------
+    # HELPERS
+    # -------------------------------------------------------------------------
+
+    @api.model
+    def _l10n_vn_edi_format_date(self, date):
+        """
+        All APIs for Sinvoice uses the same time format, being the current hour, minutes and seconds converted into
+        seconds since unix epoch, but formatting like milliseconds since unix epoch.
+        It means that the time will end in 000 for the milliseconds as they are not as of today used by the system.
+        """
+        return int(date.timestamp()) * 1000 if date else 0
+
+    @api.model
+    def _l10n_vn_edi_format_phone_number(self, number):
+        """
+        Simple helper that takes in a phone number and try to format it to fit sinvoice format.
+        SInvoice only allows digits, so we will remove any (, ), -, + characters.
+        """
+        # We first replace + by 00, then we remove all non digit characters.
+        number = number.replace('+', '00')
+        return re.sub(r'[^0-9]+', '', number)
+
+    def _l10n_vn_edi_get_sent_statuses(self):
+        self.ensure_one()
+        return {'sent', 'payment_state_to_update', 'canceled', 'adjusted', 'replaced'}
```

Comment:
I'm not sure this method is really useful since it's used only on `_l10n_vn_edi_is_sent`